### PR TITLE
[stable/traefik] Ability to add arbitrary Secrets and ConfigMaps

### LINKED ISSUE #5617: [stable/traefik] acme / DNS / gcloud / Ability to specify GCE_SERVICE_ACCOUNT_FILE
**Is this a request for help?**: NO

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

---

Setup of ACME for `gcloud` is different from the other DNS providers in that it needs a file to be accessible in the file system. Currently there is no easy way to configure `GCE_SERVICE_ACCOUNT_FILE` when using the chart.

One approach that I can verify as working is the one provided by @Azuka in the initial DNS-01 implementation here: https://github.com/kubernetes/charts/pull/1299#issuecomment-361450527. (`GCE_SERVICE_ACCOUNT_FILE` works today, `GOOGLE_APPLICATION_CREDENTIALS` is not needed anymore)

What would be a good way to implement this? @krancour @emilevauge @dtomcej @ldez

I'd gladly do an implementation for it as long as we can come to terms with what would be the best way forward.

### PR DESCRIPTION
**Pull request by jacobsvante**
**What this PR does / why we need it**:

Google Cloud DNS is configured differently from all other ACME DNS providers in that it expects a file to be available inside the container (because of how the underlying `lego` library is set up). This didn’t map well to how helm packages are expected to be provided with configuration/secrets.

To support this the ability to add any number of Secrets and ConfigMaps has been added.

~Implementation details:~
~The contents of `GCE_SERVICE_ACCOUNT_FILE_DATA` will be made available in a file inside the container and its path will be set as the value of environment variable `GCE_SERVICE_ACCOUNT_FILE`, which the [underlying ACME library](https://github.com/xenolf/lego) looks for.~

~That means that any value set manually to `GCE_SERVICE_ACCOUNT_FILE` will be ignored with this change. My general understanding of k8s/helm is quite fresh so I'm not sure if anyone's been able to use the setting at all before without resorting to something like what @Azuka suggested in https://github.com/kubernetes/charts/pull/1299#issuecomment-361450527.~

**Which issue this PR fixes**: fixes #5617

@krancour @emilevauge @dtomcej @ldez

### CODE REVIEW

File: stable/traefik/Chart.yaml
Comment (timstoop) [inline stable/traefik/Chart.yaml]: You're adding in a new feature, please bump the minor version of the chart.

**Review by timstoop — CHANGES_REQUESTED** [review]

File: stable/traefik/Chart.yaml
Comment (jacobsvante) [reply stable/traefik/Chart.yaml]: @timstoop Done

**Comment by timstoop** [comment]
Looks like another PR for traefik got accepted first. Please merge from master and check the Chart version again.

**Comment by jacobsvante** [comment]
@timstoop Merged

**Comment by timstoop** [comment]
You're still adding a new feature, so I would expect you to bump the minor version :) To 1.31.0 this time.

**Comment by jacobsvante** [comment]
Please hold with merging this PR, it looks as if the env vars aren't properly set.

**Comment by jacobsvante** [comment]
Okay, fixed the problem. Should be good to merge.

/hold cancel

**Comment by krancour** [comment]
I can easily imagine many other configuration options (if not in the present, then in the future) where a file might need to be present on the system, so I'd like to suggest a more general mechanism for handling this.

I would propose the addition of something like keys `configFiles` and `secretFiles` added to `values.yaml` and those should be keys for objects with their own keyed attributes. The values of these can be file content. i.e. Something like this:

```
configFiles:
  someArbitraryKey: |
    This is the content of some file.
secretFiles:
  someOtherArbitraryKey: |
    This is the content of some other file.
```

You can then add one new `ConfigMap` and one new `Secret` resource to the templates and those should iterate over `configFiles` and `secretFiles`, respectively to incorporate their k=v pairs.

Both of the above can be unconditionally mounted to someplace in each pod. For instance, `/var/configs` and `/var/secrets` (maybe there's a better place; I'm just using my imagination right now).

For your specific use case, this _generic_ mechanism would allow you to specify the content of the service account file and when configuring the DNS challenge to use Google Cloud as the provider, you would use the `GCE_SERVICE_ACCOUNT_FILE` as normal to reference the location of the file, which would be `/var/secrets/<key>`.

Again-- I'd favor this more generic approach because I think it would get us out of more jams in the future.

**Comment by jacobsvante** [comment]
@krancour changed as per your suggestion.

Maybe I should edit title and description to reflect the new generic functionality?

File: stable/traefik/templates/config-files.yaml
Comment (krancour) [inline stable/traefik/templates/config-files.yaml]: Willing to be corrected if I am wrong, but I am pretty sure that values in a config map aren't supposed to be base64 encoded.

File: stable/traefik/Chart.yaml
Comment (krancour) [inline stable/traefik/Chart.yaml]: If this were merged next, version should be 1.32.0.

**Comment by krancour** [comment]
Nice improvements in that last commit @jmagnusson. Thanks. See comment about base64 in config map.

File: stable/traefik/templates/config-files.yaml
Comment (jacobsvante) [reply stable/traefik/templates/config-files.yaml]: Thanks. Didn't test that one

**Comment by jacobsvante** [comment]
@krancour Removed the `base64` of configmaps. Lgtm?

**Comment by jameswinegar** [comment]
Just for prosperity. We just tested with for one of our projects. Works well.

**Comment by aroq** [comment]
Hey, does it have any chance to be merged into trunk at some point?

**Comment by jmagnusson** [comment]
Hmmm.. Temporary error perhaps?

**Comment by rosskevin** [comment]
I'm trying out this PR (it looks great, thanks!), and I'm just trying to get to the traefik dashboard successfully as a first step.  I'm receiving:
```
time="2018-08-02T21:53:51Z" level=error msg="Unable to obtain ACME certificate for domains \"traefik.mydomain.com\" detected thanks to rule \"Host:traefik.mydomain.com\" : cannot obtain certificates: acme: Error -&gt; One or more domains had a problem:\n[traefik.mydomain.com] Error presenting token: googleapi: Error 409: The resource 'entity.change.additions[0]' named '_acme-challenge.traefik.mydomain.com. (TXT)' already exists, alreadyExists\n"
```
even though that statement appears to be untrue.  I'm wondering if it is a misconfiguration on my part.    Many details in slack: https://traefik.slack.com/archives/C0CDT22PJ/p1533246103000064

My first thought is that I haven't configured properly - but I cannot identify the problem.  Values etc are in slack.  Any chance you could take a peek @jmagnusson?  I was sure to setup the service account and policy binding with `roles/dns.admin`.

**Comment by jmagnusson** [comment]
@rosskevin I think that’s the same error I’ve encountered many times, especially when multiple domains are configured. I don’t think it has anything to do with this PR but rather something with the underlying Lego library’s interaction with the Google Cloud DNS API (something with retries perhaps?).

My workaround has been to kill the traefik pod so that the acme renewal procedure will start over, which usually fixed the problem. If not you might have to fix/recreate Traefik’s acme json file.

**Comment by jameswinegar** [comment]
Bump to remove stale? @krancour any thoughts in this PR.

**Comment by jmagnusson** [comment]
I'd gladly fix the mergeability issue if you give okay to this PR

**Comment by jmagnusson** [comment]
I've bumped the chart version and signed off the commit.

@juliens @krancour @emilevauge @dtomcej @ldez any chance we could get this approved and merged now?

**Comment by jmagnusson** [comment]
@juliens Can you add /ok-to-test?

**Comment by krancour** [comment]
fyi to those asking me to look... I stepped down as a maintainer of this chart many months ago.

**Comment by nreisbeck** [comment]
@jmagnusson and chance to resolve the conflict in Charts.yaml & README.md and ping a few more maintainers for an ok to test?

I cherry picked this and have also been using it for a few months, a nice workaround for the acme-lego limitations. 👍

**Comment by aasm-moura** [comment]
Nothing guys?

**Comment by dtomcej** [comment]
@jmagnusson Sorry for the delay in review. If you can rebase/bump the chart version, we can get this merged in!

**Comment by aasm-moura** [comment]
I think the implementation is wrong!

We need to use `GOOGLE_APPLICATION_CREDENTIALS` instead of `GCE_SERVICE_ACCOUNT_FILE`  according to https://github.com/helm/charts/pull/1299#issuecomment-361450527

**Comment by jmagnusson** [comment]
@aasm-moura It does work with the underlying `lego` library. See here for relevant code in lego that uses `GCE_SERVICE_ACCOUNT_FILE`: https://github.com/xenolf/lego/blob/0e6e4807b19b5ec3bbe0c77dc7d9bb9b76c92156/providers/dns/gcloud/googlecloud.go#L53.
(Also I can confirm that the code from this PR works, I'm using it in production)

**Comment by jmagnusson** [comment]
@dtomcej Rebased on top of latest. Thanks for going forward with the merge!

**Comment by nreisbeck** [comment]
> @dtomcej Rebased on top of latest. Thanks for going forward with the merge!

@jmagnusson You're merge conflicted again on `Charts.yaml` will need another rebase

**Comment by jwheeler-gs** [comment]
If I'm reading the code in this PR correctly, it looks like the only way to have a secret mounted in is by specifying one or more values in "secretFiles". Doesn't this mean that I have to put my service account key into the values file? That's not very good secret management.

Can we set this up to also allow mounting in a secret that is already present in the cluster? This way the service account key only has to be loaded as a secret into the cluster one time and then traefik can simply use that existing secret without re-importing it from a plain text value file every time a helm upgrade is run.

**Comment by jmagnusson** [comment]
@jwheeler-gs I don't really have the use case for that personally, but if you do you can always create a PR.

**Comment by jmagnusson** [comment]
@nreisbeck Rebased on top of latest

**Comment by jmagnusson** [comment]
@dtomcej  Any chance we could get this merged in now that it's freshly rebased?

**Comment by jmagnusson** [comment]
@dtomcej Thanks!
Any idea why it's not merging? The PR has both `approved` and `lgtm` labels

**Pull request merged**
